monai/data/image_reader.py

```diff
@@ -0,0 +1,325 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional, Sequence, Tuple, Union
+
+import numpy as np
+
+from monai.data.utils import correct_nifti_header_if_necessary
+from monai.utils import ensure_tuple, optional_import
+from .utils import is_supported_format
+
+itk, _ = optional_import("itk", allow_namespace_pkg=True)
+nib, _ = optional_import("nibabel")
+
+
+class ImageReader(ABC):
+    """Abstract class to define interface APIs to load image files.
+    users need to call `read` to load image and then use `get_data`
+    to get the image data and properties from meta data.
+
+    """
+    @abstractmethod
+    def verify_suffix(self, filename: str) -> bool:
+        """
+        Verify whether the specified file or files format is supported by current reader.
+
+        Args:
+            filename: file name or a list of file names to read.
+                if a list of files, verify all the subffixes.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+
+    @abstractmethod
+    def read(self, filename: str) -> Union[Sequence[Any], Any]:
+        """
+        Read image data from specified file or files and save to `self.img`.
+        Note that it returns the raw data, so different readers return different image data type.
+
+        Args:
+            filename: file name or a list of file names to read.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def get_data(self) -> Tuple[np.ndarray, Dict]:
+        """
+        Extract data array and meta data from loaded image and return them.
+        This function must return 2 objects, first is numpy array of image data, second is dict of meta data.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+
+class ITKReader(ImageReader):
+    """
+    Load medical images based on ITK library.
+    All the supported image formats can be found:
+    https://github.com/InsightSoftwareConsortium/ITK/tree/master/Modules/IO
+
+    Args:
+        img: image to initialize the reader, this is for the usage that the image data
+            is already in memory and no need to read from file again, default is None.
+        c_order_axis_indexing: if `True`, the numpy array will have C-order indexing.
+            this is the reverse of how indices are specified in ITK, i.e. k,j,i versus i,j,k.
+            however C-order indexing is expected by most algorithms in numpy / scipy.
+
+    """
+
+    def __init__(self, img: Optional[itk.Image] = None, c_order_axis_indexing: bool = False):
+        super().__init__()
+        self.img = img
```

Comment:
I think `self.img = img` should be done in `ITKReader.read(...)` according to `ImageReader.read` docstring
```
def read(self, data: Any):
     if data is string-like filename:
           self.img = self.read_by_file_name(data)
     if data is string like directory name:
           self.img = self.read_by_dir_name(data)
     if data is itk img object:
           self.img = data
```
The current constructor is a specific case of `ITKReader.read` which is confusing...

Reply:
OK, make sense to move to `read`.
Thanks.